interface/main/messages/templates/linked_documents.php

```diff
@@ -0,0 +1,237 @@
+               href='javascript:void(0);'
+               onClick="gotoReport(<?php echo attr_js($record['documentId']); ?>, '<?php echo attr_js($record['pname']); ?>', '<?php echo attr_js($record['pid']); ?>','<?php echo attr_js($record['pubpid'] ?? $record['pid']); ?>','<?php echo attr_js($record['DOB']); ?>');">
+                <?php echo text($record['title']); ?>
+            </a>
+        <?php endif; ?>
+        <?php if ($record['requiresValidation']) : ?>
+            <span class="validation-line">
+                <?php echo text($record['title']); ?>
+                <span class="text-info"><?php echo xlt("Validating document for errors"); ?>...</span>
+                <span class="spinner-border-sm spinner-border" role="status">
+                    <span class="sr-only"><?php echo xlt("Validating document for errors"); ?>...</span>
+                </span>
+            </span>
+            <span class="validation-failed text-danger d-none">
+                <?php echo text($record['title']); ?> -
+                <?php echo xlt("Failed to retrieve validation results from server"); ?>
+            </span>
+        <span class="validation-totals validation-totals-success badge badge-pill badge-success p-2 d-none">
+            <?php echo xlt("Errors"); ?> <span class="validation-totals-count">0</span>
+        </span>
+            <span class="validation-totals validation-totals-failed badge badge-pill badge-danger p-2 d-none">
+            <?php echo xlt("Errors"); ?> <span class="validation-totals-count">0</span>
+        </span>
+        <?php endif; ?>
+        <?php if ($record['isCda']) : ?>
+        <details class="validation-report-errors d-none">
+            <summary><?php echo xlt("Validation report"); ?></summary>
+            <div class="validation-report-errors-container">
+            </div>
+        </details>
+    <?php endif; ?>
+    </div>
+</div>
+    <?php endforeach;
+} catch (Exception $exception) {
+    // if twig throws any exceptions we want to log it.
+    (new SystemLogger())->errorLogCaller($exception->getMessage(), ['trace' => $exception->getTraceAsString()]);
+}
+?>
+
+<script>
+    function startDocumentValidation() {
+        top.restoreSession();
+        // document validation can take a long time especially if it is happening externally...
+        // we are going to do an async process to go out and validate the document before we allow the document to be
+        // loaded.
+        let records = document.querySelectorAll(".messages-document-validate");
+        records.forEach(function(validateRecord) {
+
+            // now we need to make an ajax async request to the server with the document id
+            let docId = validateRecord.dataset['doc'];
+
+            window.fetch('<?php
+                echo $GLOBALS['webroot'] . "/library/ajax/messages/validate_messages_document_ajax.php?csrf=" . urlencode(CsrfUtils::collectCsrfToken());
+            ?>' + "&doc=" + encodeURIComponent(docId))
```

Comment:
whoohooo, get to use the rarely used js_url() function here:
```
echo $GLOBALS['webroot'] . "/library/ajax/messages/validate_messages_document_ajax.php?csrf="; ?>' + <?php echo js_url(CsrfUtils::collectCsrfToken()); ?> + "&doc=" + encodeURIComponent(docId))
```

Reply:
this was a lot of work to change this around to use the js_url.  What a pain, but I think I got it covered now.